File: imgtomatrix.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <chrono>

using namespace cv;
using namespace std;

// Function to extract grid and return board configuration
tuple<vector<vector<int>>, map<string, Scalar>, int, tuple<int, int>, tuple<int, int>> extract_grid(const string& file_name) {
    auto start = chrono::high_resolution_clock::now();

    // Read the input image
    Mat original = imread(file_name);
    if (original.empty()) {
        cerr << "Error: Image not found!" << endl;
        return {};
    }
    
    // Save the original image for reference
    imwrite("solution/original.png", original);

    // Convert the image to grayscale for easier contour detection
    Mat gray;
    cvtColor(original, gray, COLOR_BGR2GRAY);

    // Find contours in the grayscale image
    vector<vector<Point>> contours;
    findContours(gray, contours, RETR_TREE, CHAIN_APPROX_NONE);

    // Sort contours by area in descending order
    sort(contours.begin(), contours.end(), [](const vector<Point>& a, const vector<Point>& b) {
        return contourArea(a) > contourArea(b);
    });

    // Extract the bounding box of the puzzle grid (usually the second largest contour)
    Rect bounding_box = boundingRect(contours[1]);

    // Crop the grid area from the original image
    Mat grid = original(bounding_box);

    // Save the cropped grid image
    imwrite("solution/grid.png", grid);

    // Convert the cropped grid to grayscale
    cvtColor(grid, gray, COLOR_BGR2GRAY);
    imwrite("solution/gray-grid.png", gray);

    // Find contours again within the cropped grayscale grid
    Mat thresh;
    threshold(gray, thresh, 127, 255, THRESH_BINARY);
    findContours(thresh, contours, RETR_TREE, CHAIN_APPROX_NONE);

    // Sort contours by area
    sort(contours.begin(), contours.end(), [](const vector<Point>& a, const vector<Point>& b) {
        return contourArea(a) < contourArea(b);
    });

    // Calculate the total number of cells and the grid size
    int total_cells = contours.size() - 2;
    if (total_cells <= 0) {
        cerr << "No valid cells detected! Aborting" << endl;
        return {};
    }

    int grid_size = static_cast<int>(sqrt(total_cells));
    if (total_cells != grid_size * grid_size) {
        cerr << "Unable to detect full grid! Aborting" << endl;
        return {};
    }

    // Calculate the width and height of each cell
    int cell_width = bounding_box.width / grid_size;
    int cell_height = bounding_box.height / grid_size;

    // Initialize variables to store colors and board configuration
    map<string, Scalar> color_map;
    map<int, Scalar> reverse_color_map;
    int color_index = 1;
    vector<vector<int>> board;

    // Iterate through each cell in the grid to extract colors
    for (int i = 0; i < grid_size; ++i) {
        vector<int> row;
        for (int j = 0; j < grid_size; ++j) {
            // Calculate cell coordinates with padding
            int padding = 15;
            int cell_x = j * cell_width;
            int cell_y = i * cell_height;
            Rect cell_rect(cell_x + padding, cell_y + padding, cell_width - 2 * padding, cell_height - 2 * padding);
            Mat cell = grid(cell_rect);

            // Get the average color of the cell
            Scalar avg_color = mean(cell);
            string color_key = to_string(int(avg_color[0])) + "," + to_string(int(avg_color[1])) + "," + to_string(int(avg_color[2]));

            // Map each unique color to an index
            if (color_map.find(color_key) == color_map.end()) {
                color_map[color_key] = avg_color;
                reverse_color_map[color_index] = avg_color;
                row.push_back(color_index++);
            } else {
                row.push_back(distance(color_map.begin(), color_map.find(color_key)) + 1);
            }
        }
        board.push_back(row);
    }

    // Check if the number of detected colors matches the grid size; abort if mismatched
    if (color_map.size() != grid_size) {
        cerr << "Too many colors detected! Aborting" << endl;
        return {};
    }

    // Print the C++ board format for debugging
    cout << "C++ Board Format:" << endl;
    cout << "vector<vector<int>> board = {" << endl;
    for (size_t i = 0; i < board.size(); ++i) {
        cout << "    {";
        for (size_t j = 0; j < board[i].size(); ++j) {
            cout << board[i][j];
            if (j < board[i].size() - 1)
                cout << ", ";
        }
        cout << "}";
        if (i < board.size() - 1)
            cout << ",";
        cout << endl;
    }
    cout << "};" << endl;

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end - start;
    cout << "Time taken: " << elapsed.count() << " seconds" << endl;

    return make_tuple(board, color_map, grid_size, make_tuple(bounding_box.width, bounding_box.height), make_tuple(cell_width, cell_height));
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <image_filename>" << endl;
        return 1;
    }

    string file_name = argv[1];
    auto result = extract_grid(file_name);

    if (!get<0>(result).empty()) {
        vector<vector<int>> board = get<0>(result);
        auto color_map = get<1>(result);
        int grid_size = get<2>(result);
        auto dimensions = get<3>(result);
        auto cell_dimensions = get<4>(result);

        cout << "Grid size: " << grid_size << "x" << grid_size << endl;
        cout << "Board representation:" << endl;
        for (const auto& row : board) {
            for (const auto& cell : row) {
                cout << cell << " ";
            }
            cout << endl;
        }
    }

    return 0;
}

File: src/imgtomatrix.py
import cv2 as cv
import numpy as np
import math
import time

def extract_grid(file_name):
    start = time.perf_counter()

    # Read the input image
    original = cv.imread(file_name)
    # Save the original image for reference
    cv.imwrite("solution/original.png", original)

    # Convert the image to grayscale for easier contour detection
    gray = cv.cvtColor(original, cv.COLOR_BGR2GRAY)

    # Find contours in the grayscale image
    contours, _ = cv.findContours(gray, cv.RETR_TREE, cv.CHAIN_APPROX_NONE)
    # Sort contours by area in descending order
    contours = sorted(contours, key=cv.contourArea, reverse=True)

    # Extract the bounding box of the puzzle grid (usually the second largest contour)
    x, y, w, h = cv.boundingRect(contours[1])

    # Crop the grid area from the original image
    grid = original[y:y+h, x:x+w]
    # Save the cropped grid image
    cv.imwrite("solution/grid.png", grid)

    # Convert the cropped grid to grayscale
    gray = cv.cvtColor(grid, cv.COLOR_BGR2GRAY)
    # Save the grayscale grid image
    cv.imwrite("solution/gray-grid.png", gray)

    # Find contours again within the cropped grayscale grid
    # Apply threshold to ensure better contour detection
    _, thresh = cv.threshold(gray, 127, 255, cv.THRESH_BINARY)
    contours, _ = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_NONE)
    # Sort contours by area
    contours = sorted(contours, key=cv.contourArea)

    # Calculate the total number of cells and the grid size
    total_cells = len(contours) - 2
    if total_cells <= 0:
        print("No valid cells detected! Aborting")
        return None
    
    grid_size = int(math.sqrt(total_cells))
    if total_cells != grid_size**2:
        print("Unable to detect full grid! Aborting")
    # Initialize variables to store board configuration
    # Calculate the width and height of each cell
    cell_width = w // grid_size
    cell_height = h // grid_size

    # Initialize variables to store colors and board configuration
    colors = []
    board = []
    color_index = 1
    color_map = {}
    reverse_color_map = {}
    padding = 10

    # Iterate through each cell in the grid to extract colors
    for i in range(grid_size):
        row = []
        for j in range(grid_size):
            # Calculate cell coordinates with padding
            cell_x = j * cell_width
            cell_y = i * cell_height
            padding = 15
            cell = grid[cell_y+padding:cell_y+cell_height-padding, cell_x+padding:cell_x+cell_width-padding]
            
            # Get the average color of the cell
            avg_color = cell.mean(axis=0).mean(axis=0)
            avg_color = avg_color.astype(int)
            avg_color = tuple(avg_color)
            
            # Map each unique color to an index
            if avg_color not in color_map:
                color_map[avg_color] = str(color_index)
                reverse_color_map[str(color_index)] = avg_color
                color_index += 1
            row.append(color_map[avg_color])
            
        # Append the row to the board configuration
        board.append(row)
                
    # Check if the number of detected colors matches the grid size; abort if mismatched
    if len(color_map) != grid_size:
        print("Too many colors detected! Aborting")
        return None

    # Convert the string representation to integers for C++ compatibility
    cpp_board = []
    for row in board:
        cpp_row = [int(cell) for cell in row]
        cpp_board.append(cpp_row)
    
    print("C++ Board Format:")
    print("vector<vector<int>> board = {")
    for i, row in enumerate(cpp_board):
        print(f"    {{{', '.join(map(str, row))}}}", end="")
        if i < len(cpp_board) - 1:
            print(",")
        else:
            print("")
    print("};")

    
    end = time.perf_counter()
    elapsed = end - start
    print(f'Time taken: {elapsed:.6f} seconds')
    
    return board, reverse_color_map, grid_size, (w, h), (cell_width, cell_height)

# Example usage
if __name__ == "__main__":
    result = extract_grid("puzzle.png")
    if result:
        board, color_map, grid_size, dimensions, cell_dimensions = result
        print(f"Grid size: {grid_size}x{grid_size}")
        print("Board representation:")
        for row in board:
            print(row)
